### DISCUSSION TITLE
Can I make a "higher order" machine that transitions based on the state of submachines?

### DISCUSSION THREAD

**Discussion by ez-alexfrazer** [statelyai/xstate#2074] [Q&A]
Here's the concept: imagine I have a machine responsible for validation of some input. Here's a quick model of what it would look like.

![image](https://user-images.githubusercontent.com/77698413/114072245-086c3380-9870-11eb-9e19-65710ec3bf69.png)

This works perfectly well. It will allow me to update the input and check its validity in my use case. It also won't allow me to commit until the input has been determined to be valid.

Now here's what I'm trying to accomplish: I want a parent machine that can "register" n number of these machines and transition state based on the state of these machines. i.e., when _all_ of the machines have registered as valid, then the machine that spawns them will transition automatically to a valid state as well.

I looked a bit into doing this, and I found [Actors](https://xstate.js.org/docs/guides/actors.html#actor-api). I tried to work with it a bit and do something like this:

```
const fieldMachine = Machine({/** ... */});

const machine = Machine({
  id: 'validation-machine',
  initial: 'active',
  context: {
    fields: [],
  },
  states: {
    active: {
      on: {
        REGISTER: {
          actions: assign({
            fields: (ctx, event) => [
              ...ctx.fields,
              spawn(fieldMachine, event.name),
            ]
          }),
        }
      }
    }
  }
});
```

I'm not sure if I'm on the right path, but has anyone tried to do something like this before?

**Comment by ezhikov** [comment]
I think, that you  are going in right direction, but you should be careful about how chatty your children are and what libraries you are connecting to. For example, in case of react, passing chatty services down the tree may result in high amount of renders when component is rendered twice - first from local state update (in `useService` hook) and then from parent state update (when message will reach the parent in next tick). I had this  problem and **fastest** solution was to [manually trigger `"xstate.update"` in parent](https://github.com/davidkpiano/xstate/discussions/1614) when needed, without any work from children. While it works, I believe, this is a bad practice, since it is internal event and it's generally not very good idea to meddle with internal stuff. It should be sent from child to parent via specific action creator when needed.

So, by default you will have to [manually notify parent](https://xstate.js.org/docs/guides/actors.html#sending-updates) machine about changes in children, so it is not "parent watches child", but "child cares to phone home". I would probably keep every field value and validations status in one place now, but we don't have instant validation - only when user tries to send form. If your field will emit only two events to parent (when state changes from `invalid` to `valid` and vice versa, I think you will be fine.

**Comment by fraserdrops** [comment]
Yes, I think you're on the right track spawning actors. One of the giveaways for me is when you describe being able to register machines dynamically, which is exactly what `spawn` is for. 

As @ezhikov said, with spawning actors they will need to tell the parent (`sendParent` is handy here) when they want the parent to know something; so in this case, you'd probably want the child to message the parent when it reaches the `valid` state.

When the parent is getting messages from the spawned actors, it would need to keep track of which children are valid and which aren't. At first, this may seem like duplicated effort, because the child itself already knows what state its in. But with the actor model, the way for one actor to know something is for it to be explicitly told through a message. The parent doesn't need to keep all the internal state of the child, but it will need to store which children are valid and which aren't somewhere in context - something like the code below

```const fieldMachine = Machine({/** ... */});

const machine = Machine({
  id: 'validation-machine',
  initial: 'active',
  context: {
    fields: [],
  },
  states: {
    active: {
      on: {
        REGISTER: {
          actions: assign({
            fields: (ctx, event) => {
              ...ctx.fields,
              [event.name]: {
                ref:   spawn(fieldMachine, event.name),
                isValid: false
              }
            }
          }),
        }
        UPDATE_FROM_CHILD: {
          actions: assign({
            fields: (ctx, event) => [
              ...ctx.fields,
              [event.name]: {
                ref:   spawn(fieldMachine, event.name),
                isValid: event.isValid
              }
            ]
          }),
        }
      }
    }
  }
});
```

After that, you'll need to check after each update whether all the children are valid and then transition.

Hope that helps! In general, the pattern of the parent keeping a bit of information about its children and the child letting the parent know  when this data changes seems to come up a bit when using actors, at least in my experience.
